### PR TITLE
override consul-template blacklist configuration

### PR DESCRIPTION
**Pull request by tgross**
By default consul-template allows any template rendering function it supports to be used. This PR drops the `plugin` and `file` functions by default for Nomad templates. Includes:
- Agent-level configuration option for enabling rendering w/ these functions in case a cluster operator wants to permit them.
- Pulls in hashicorp/consul-template#1243 and hashicorp/consul-template#1246

### CODE REVIEW

**Comment by tgross** [comment]
(I seemed to have screwed something up in the vendoring of the updated consul-template here so it's failing CI. Let me fix that 😊 )

File: client/allocrunner/taskrunner/template/template.go [hunk 1]
```diff
@@ -574,6 +574,13 @@ func parseTemplateConfigs(config *TaskTemplateManagerConfig) (map[ctconf.Templat
 		ct.LeftDelim = &tmpl.LeftDelim
 		ct.RightDelim = &tmpl.RightDelim
 
+		// By default we pass a blacklist of functions to prevent
+		// task operators from bypassing client-task isolation.
+		// This protection can be disabled by the client config.
+		if !config.ClientConfig.EnableInsecureTemplateFunctions {
```
Comment (cgbaker) [inline client/allocrunner/taskrunner/template/template.go]: this line and the change to the documentation suggest that this is supposed to be in the client config, but the change to the parsing show it in the top-level config. as such, right now it's not possible to change this to a non-default value of `true`.

File: command/agent/config.go [hunk 2]
```diff
@@ -108,6 +108,10 @@ type Config struct {
 	// for security bulletins
 	DisableAnonymousSignature bool `hcl:"disable_anonymous_signature"`
 
+	// EnableInsecureTemplateFunctions enables templates to include functions
+	// that are unsafe because they expose information from the client host.
+	EnableInsecureTemplateFunctions bool `hcl:"enable_insecure_template_functions"`
```
Comment (cgbaker) [inline command/agent/config.go]: this needs to move below into `ClientConfig`

**Comment by jippi** [comment]
Can we make this a feature flag? Especially for `file`, there is plenty of use-cases in the world where a template is kinda like this

```hcl
template {
    destination = "pki.json"
    data = <<<EOF
{{ with secret "pki/issue/my-domain-dot-com" "common_name=foo.example.com" }}
{{ .Data | toJSON }}{{ end }}
    EOF;
}

template {
    destination = "certificate.tls"
    data = <<<EOF
{{ file "pki.json" | parseJSON | .certificate }}
   EOF;
}

template {
    destination = "stuff.tls"
    data = <<<EOF
{{ file "pki.json" | parseJSON | .other_field }}
   EOF;
}
```

File: client/allocrunner/taskrunner/template/template.go (hunk 1 above)
Comment (jippi) [inline client/allocrunner/taskrunner/template/template.go]: would like to be able to disable `plugin` but keep `file` around - maybe expose a []string for the functions I want to blacklist manually as well?

Comment (tgross) [reply client/allocrunner/taskrunner/template/template.go]: Ah yes, thanks for catching that.

**Comment by tgross** [comment]
@cgbaker I'm still working on this but in the meantime I believe I've fixed the configuration problem at this point. If I build the current PR as `make dev` and use the following configuration file:

```
bind_addr = "0.0.0.0"
data_dir = "/tmp/nomad"
log_level = "DEBUG"

server {
  enabled = true
  bootstrap_expect = 1
}

client {
  enabled = true
  enable_insecure_template_functions = true
}
```

I can confirm the correct behavior when I toggle our boolean value.

**Comment by tgross** [comment]
> Can we make this a feature flag?

Hello @jippi, thanks for your feedback on this! When I worked on this I had the same concern you do here that some operators would want to be able to be more selective about what functions they disallow. There’s a bit of a trade-off to make between configuration complexity and secure defaults. That being said, after seeing your example and having some other discussions I think I'm leaning towards having a more fine-grained control.

> I wonder if a better long term fix would be to provide a list of folders where file would be allowed to read from, and default to /local - since it’s very all-or-nothing breaking a safe-common pattern without being able to be protected from the root cause that made you implement this whitelisting of functions

I would definitely like to see this myself! If we could sandbox consul-template to the same context as the task we wouldn’t really need to worry about the `file` function at all. But existing running jobs will be broken if we change the behavior, so we'll need to consider how to do this. I'm chatting with one of the consul-template developers about how the best way forward on that.

**Comment by tgross** [comment]
Just pushed a set of commits leveraging the work done in this draft PR https://github.com/hashicorp/consul-template/pull/1249 for discussion. Demo of what this looks like:

---

default.hcl
```hcl
bind_addr = "0.0.0.0"
data_dir = "/tmp/nomad"
log_level = "DEBUG"

server {
  enabled = true
  bootstrap_expect = 1
}

client {
  enabled = true
  template {
    function_blacklist = ["plugin"]
    disable_sandbox = false
  }
}
```

file-leak.nomad
```hcl
job "example" {
  datacenters = ["dc1"]

  type = "service"

  group "cache" {
    count = 1

    task "redis" {
      driver = "docker"
      config {
        image = "redis:3.2"
        port_map {
          db = 6379
        }
      }
      resources {
        cpu    = 500 # 500 MHz
        memory = 256 # 256MB
        network {
          mbits = 10
          port "db" {}
        }
      }


      template {
        data          = "{{ file \"/etc/passwd\" }}"
        destination   = "local/file.yml"
      }
    }
  }
}
```

output:
```
    2019/08/07 11:24:04.940971 [ERR] (runner) watcher reported error: file(/tmp/nomad/alloc/a6dc1786-2e98-3fcf-7029-a24a97ab0ec6/redis/etc/passwd): stat /tmp/nomad/alloc/a6dc1786-2e98-3fcf-7029-a24a97ab0ec6/redis/etc/passwd: no such file or directory
```

**Comment by tgross** [comment]
We're getting a _lot_ of timeouts on Travis builds here. I've opened https://github.com/hashicorp/nomad/pull/6094 to help mitigate some of that.

**Comment by tgross** [comment]
@cgbaker @jippi I think I'm happy with where this has landed at this point.  The configuration values and their defaults are:

```hcl
client {
  template {
    function_blacklist = ["plugin"]
    disable_file_sandbox = false
  }
}
```

Although I know @jippi mentioned wanting to sandbox tasks to specific directories outside the task directory, we're going to punt on that for now and have a flag to disable the sandbox. We can return to this in later work and see if it makes sense to include, but we need the all-or-nothing flag for backwards compatibility. For now, this new configuration gives folks a bunch of different options depending on the level of protection they need:

- `function_blacklist = ["file", "plugin"]`: disable the `file` function entirely
- `disable_file_sandbox = true`: allow the `file` function and allow it access to arbitrary files on the client
- `disable_file_sandbox = false` + `docker { volume ...`: use a docker bind mount to include specific files from the client into the task directory sandbox where the `file` function can read them.
- `disable_file_sandbox = false` + [docker volumes `enabled=false`](https://www.nomadproject.io/docs/drivers/docker.html#enabled): disallow docker bind mounts and sandbox the `file` function to the task directory.

What do we think?

**Comment by tgross** [comment]
Rebased this on master to pick up #6095 and hopefully bring test run times down.

File: website/source/docs/configuration/client.html.md [hunk 3]
```diff
@@ -329,6 +333,18 @@ see the [drivers documentation](/docs/drivers/index.html).
   reserve on all fingerprinted network devices. Ranges can be specified by using
   a hyphen separated the two inclusive ends.
 
+
+### `template` Parameters
+
+- `function_blacklist` `([]string: ["plugin"])` - Specifies a list of template
+  rendering functions that should be disallowed in job specs because they can
+  leak information from the client host to templates.
```
Comment (schmichael) [inline website/source/docs/configuration/client.html.md]: "leak information" is too weak to describe `plugin`, so maybe we should expand with:

> By default the `plugin` function is disallowed as it allows running arbitrary commands on the host as root (unless Nomad is configured to run as a non-root user).

Comment (tgross) [reply website/source/docs/configuration/client.html.md]: Done.

**Comment by tgross** [comment]
Looks like we can avoid breaking existing uses of `NOMAD_TASK_DIR` in the `file` parameter of templates if we use the patch to consul-template in https://github.com/hashicorp/consul-template/pull/1254.

File: vendor/vendor.json [hunk 4]
```diff
@@ -6,6 +6,7 @@
 		{"path":"github.com/Azure/azure-sdk-for-go/version","checksumSHA1":"FAw+h8wS2QiQEIVC3z/8R2q1bvY=","revision":"767429fcb996dad413936d682c28301e6739bade","revisionTime":"2018-05-01T22:35:11Z"},
 		{"path":"github.com/Azure/go-ansiterm","checksumSHA1":"9NFR6RG8H2fNyKHscGmuGLQhRm4=","revision":"d6e3b3328b783f23731bc4d058875b0371ff8109","revisionTime":"2017-09-29T23:40:23Z","version":"master","versionExact":"master"},
 		{"path":"github.com/Azure/go-ansiterm/winterm","checksumSHA1":"3/UphB+6Hbx5otA4PjFjvObT+L4=","revision":"d6e3b3328b783f23731bc4d058875b0371ff8109","revisionTime":"2017-09-29T23:40:23Z","version":"master","versionExact":"master"},
+		{"path":"github.com/BurntSushi/toml","checksumSHA1":"Pc2ORQp+VY3Un/dkh4QwLC7R6lE=","revision":"3012a1dbe2e4bd1391d42b32f0577cb7bbc7f005","revisionTime":"2018-08-15T10:47:33Z"},
```
Comment (notnoop) [inline vendor/vendor.json]: Probably should remove https://github.com/hashicorp/nomad/blob/b61ced1c802fe6ca2a389a75fc2f6c612939eefa/vendor/vendor.json#L57 - it seems CT is the only library references toml library.

Comment (tgross) [reply vendor/vendor.json]: Fixed.

**Comment by tgross** [comment]
This work has been cherry-picked into the `release-095` branch.

**Pull request merged**